File: h/按之字形顺序打印二叉树.java
/*
请实现一个函数按照之字形打印二叉树，即第一行按照从左到右的顺序打印，
第二层按照从右至左的顺序打印，第三行按照从左到右的顺序打印，
其他行以此类推
*/
import java.util.ArrayList;

/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
import java.util.*;
public class Solution {
    public ArrayList<ArrayList<Integer> > Print(TreeNode pRoot) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(pRoot == null){
            return res;
        }
        ArrayList<Integer> list;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        int rows = 1;
        queue.add(pRoot);
        while (!queue.isEmpty()) {
            list = new ArrayList();
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode t = queue.poll();
                if (rows % 2 == 0){
                    list.add(0, t.val);
                } else {
                    list.add(t.val);
                }
                if (t.left != null) {
                    queue.offer(t.left);
                }
                if (t.right != null) {
                    queue.offer(t.right);
                }
            }
            res.add(list);
            rows++;
        }
        return res;
    }
}

File: z/按之字形顺序打印二叉树.cc
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
 public:
  vector<vector<int>> Print(TreeNode* pRoot) {
    vector<vector<int>> res;
    if (pRoot == NULL) {
      return res;
    }
    queue<TreeNode*> q;
    q.push(pRoot);
    bool flag = true;
    TreeNode* iter;
    while (!q.empty()) {
      int len = q.size();
      vector<int> level(len);
      for (int i = 0; i < len; ++i) {
        iter = q.front();
        int index = flag ? i : len - 1 - i;
        level[index] = iter->val;
        q.pop();
        if (iter->left != NULL) {
          q.push(iter->left);
        }
        if (iter->right != NULL) {
          q.push(iter->right);
        }
      }
      flag = !flag;
      res.push_back(level);
    }
    return res;
  }
};
